### PR TITLE
Create non-blocking broadcaster helper and use it to manage Coordinator state notifications

### PR DESCRIPTION
**Pull request by faec**
Create a new helper component, `Broadcaster`, which broadcasts changes in an observed value to a variable list of subscribers while providing various performance / non-blocking guarantees (see https://github.com/elastic/elastic-agent/issues/2819). Replace `Coordinator`'s current [racey state broadcasting](https://github.com/elastic/elastic-agent/issues/2735) with calls to this component.

While the core functionality is provided in `Broadcaster`, this also required significant changes to `Coordinator` itself, since we need to generate a reliable stream of discrete state values to pass it as input.

The biggest design change to enable this was to move all state changes into the `Coordinator` goroutine, so there's a single source of truth for the value and ordering of every change to `Coordinator.state`. (Most were already there, but there were some exceptions.) This allowed some important simplifications, like removing the three mutexes that were previously needed to check any state value. In exchange, some state changes previously run externally now need to go through `Coordinator` -- e.g. `SetOverrideState` now sends to an internal channel in `Coordinator`, which applies the change synchronously on its main goroutine, instead of locking the relevant mutex and broadcasting it from an external goroutine.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] ~~I have made corresponding changes to the documentation~~
- [ ] ~~I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] ~~I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)~~
- [ ] ~~I have added an integration test or an E2E test~~

## Related issues

- Closes https://github.com/elastic/elastic-agent/issues/2735
- Closes https://github.com/elastic/elastic-agent/issues/2819
- Closes https://github.com/elastic/elastic-agent/issues/2868

### CODE REVIEW

**Comment by faec** [comment]
I'm marking this ready for review since I'm about to be gone for a 4-day weekend and the core implementation is done / ready for comments. I still need to add a bunch more tests before merging (see https://github.com/elastic/elastic-agent/issues/2868) but I've added a fair number already and they all pass (except for some intermittent CI failures on known-flaky preexisting tests), so I think it's ready for comments while I finish the remaining tests in parallel.

File: internal/pkg/agent/application/coordinator/coordinator_state.go
Comment (blakerouse) [inline internal/pkg/agent/application/coordinator/coordinator_state.go:108]: Yes this is needed. Without this it would be impossible to get the reason a unit was stopped. It would just be removed without having a reason reported in the state change. Think of the case that where another application wants to monitor and report the state changes of the Elastic Agent it would always miss this state and message.

File: pkg/control/v2/server/server.go
Comment (blakerouse) [inline pkg/control/v2/server/server.go:142]: +1 to exposing it. I believe it is completely possible that some subscribers only want the latest.

File: pkg/utils/broadcaster/broadcaster.go
Comment (blakerouse) [inline pkg/utils/broadcaster/broadcaster.go:190]: Nice! Really like that the final values can be read before shutdown.

File: pkg/utils/broadcaster/broadcaster.go
Comment (blakerouse) [inline pkg/utils/broadcaster/broadcaster.go:227]: Nice!

**Review by blakerouse — APPROVED** [review]
This defiantly needs another set of eyes, but it looks really good. Has all the items that we discussed and the documentation provided in the broadcaster is top-notch!

I do think we should expose the ability for a subscriber over the GRPC to set the buffer amount. Can be done in a followup, this change is already rather large.

File: internal/pkg/agent/application/coordinator/coordinator_state.go
Comment (faec) [reply internal/pkg/agent/application/coordinator/coordinator_state.go:108]: 👍 I think long run it's worth considering other options (like the persistent record of termination states we discussed elsewhere), but I'm happy to leave that for followup conversations, this change by itself will give us a lot more breathing room to make those choices gradually so I'll remove the TODO from this PR

**Comment by cmacknz** [comment]
Tried to run the tests locally and immediately got a failure:

```
❯ go test -race ./pkg/utils/broadcaster/...
--- FAIL: TestUnbufferedSubscriberReceivesMostRecentValue (0.00s)
    broadcaster_test.go:119:
                Error Trace:    /Users/cmackenzie/go/src/github.com/elastic/elastic-agent/pkg/utils/broadcaster/broadcaster_test.go:119
                Error:          Not equal:
                                expected: 4046
                                actual  : 4042
                Test:           TestUnbufferedSubscriberReceivesMostRecentValue
                Messages:       listener should receive the final value
FAIL
FAIL    github.com/elastic/elastic-agent/pkg/utils/broadcaster  0.429s
FAIL
```

This test seems to be flaky, running `while go test -count 1 -race ./pkg/utils/broadcaster/...; do :; done
` to took 36s of continuous tests to reproduce.

**Comment by cmacknz** [comment]
👍 This is very well commented, the comments answer pretty much all of my questions. Thank you!

**Comment by faec** [comment]
> Tried to run the tests locally and immediately got a failure

Thanks for the info! I hadn't seen that one yet and will investigate next week... it definitely needs more thorough testing in several areas, I'm happy with how it's shaping up but I do not personally trust it yet :-)

**Comment by faec** [comment]
...ok actually I spent a few more minutes looking at that test and I bet I know the cause. The test itself has a race.

Because Broadcaster's input channel is buffered, it's possible for the final input writes and the read to arrive simultaneously in the Broadcaster loop. (This is the tradeoff when making the input buffered instead of synchronous: it means slightly less scheduler churn, but immediately adjacent write-reads are not strictly guaranteed to match. The owner routine itself is always potentially one scheduler interrupt ahead of its subscribers.)

It would be easy to "fix" this particular race in Broadcaster itself -- just make its input channel unbuffered -- but I'm inclined to err on the side of a maximally responsive Coordinator, so my plan (next week for real this time) is to leave the buffer and fix the test by using the one-iteration-at-a-time model. Should probably call this out in the docs though (and maybe make the input buffer configurable so callers can get absolute causal behavior if they really need it.)

**Review by cmacknz — APPROVED** [review]
Ran the coordinator tests for 8m continuously without a failure, thanks!

File: pkg/utils/broadcaster/broadcaster_test.go
Comment (faec) [reply pkg/utils/broadcaster/broadcaster_test.go]: Thank you, fixed!

**Comment by faec** [comment]
I've finished all the new tests I'm expecting to write -- `Broadcaster` now has 98.8% unit test coverage (and I'm glad I did it, I kept finding issues at each stage), and `Coordinator` went from 66.5% to 74.4%, with a lot of the new tests covering much more fine-grained behavior than the previous ones. The remaining gaps are mostly related to manager/upgrade states that are still hard to test, and will require followup in other packages.

The last thing I'm definitely planning before it's ready for merge is a rewrite of the preexisting `TestDiagnosticHooks`, which is currently occasionally flaky and enormous (~600 hand-written lines + 1200 of auto-generated mocks)... using the new tools should make it much smaller and more consistent.

**Comment by cmacknz** [comment]
The new test cases all LGTM, looking forward to the diagnostic test refactor.

If https://github.com/elastic/elastic-agent/pull/2928 to run the integration tests on PRs isn't merged before you are done, trigger the integration tests manually from this branch before merging.

**Comment by faec** [comment]
I ended up creating two new internal Coordinator variables to facilitate the diagnostics tests, one for the derived config (AST plus variable substitutions) and one for the component model sent to the runtime manager. It doesn't change any behavior but it avoids having to call some heavy-duty internal helpers from the diagnostics, which in turn avoids having to prepare/mock those internals from the diagnostic tests. (I was planning to add these anyway as part of https://github.com/elastic/elastic-agent/issues/2852 but it simplified this final step a lot.)

Out of time for today but I expect to finish in the morning, just a few short tests left 🤞 and the diagnostics test refactor has pushed my line count dramatically into the negative :-)

**Comment by faec** [comment]
The reworked diagnostic tests turned up a bug: https://github.com/elastic/elastic-agent/issues/2940

**Comment by faec** [comment]
All tests now written and passing locally, subject to CI issues it's ready for a final look.

File: internal/pkg/agent/application/coordinator/diagnostics_test.go
Comment (pchila) [inline internal/pkg/agent/application/coordinator/diagnostics_test.go:1]: Nitpicking here: the previous version of the test was intended for having a certain (realistic) configuration in input and verifying that we generated the full set of diagnostic files with the correct information given that config.
The idea being that, if we found some bugs or some tricky configurations we could add the config here as taken from fleet and have a unit test with that.
The tests now use simplified bits of config and test 1 hook at the time for the specific bits of config.
Each approach has its own merits (the one in this PR has brevity and simplicity on its side) but here we are losing the possibility of testing the whole of the coordinator diagnostic output given certain policies/configurations.

I am not asking to change it back, I just want us to be aware of that this change implies

**Review by pchila — APPROVED** [review]

File: internal/pkg/agent/application/coordinator/diagnostics_test.go
Comment (faec) [reply internal/pkg/agent/application/coordinator/diagnostics_test.go:1]: We chatted about this a bunch offline, to summarize my perspective:
- The existing test covered a lot of ground but (flakiness aside) there was no way to tell if it was doing the right thing, since no human could audit 10K lines of golden files
- Checking internal state through the diagnostics interface, where it gets serialized back and forth to YAML, and we need to do complicated sanitization to even tell if there is a match, adds complication that is completely orthogonal to the behavior being tested -- if we want to test internal values, we can write tests that look at them directly.
- Many conditions that were previously only covered by the diagnostics test are now also covered directly in the unit tests (and the new diagnostics tests have already caught an error that was previously missed)
- Setting up "realistic" configurations can only go so far when all except one object is mocked -- this should really be the domain of the new integration/e2e tests. A complicated config with only one real component doesn't buy much... a lot of these structures are effectively opaque pointers from Coordinator's perspective, and the important part is that we send them to the right places.

All that said, we should definitely continue to expand test coverage and find ways to verify more realistic scenarios :-)

Comment (faec) [reply internal/pkg/agent/application/coordinator/diagnostics_test.go:1]: (Related followup that I think both @pchila and I agree on: the extreme verbosity of our diagnostics often isn't buying us much. I don't know anyone who has used those really enormous files productively in an SDH... the important things are the logs and a few key fields, and sometimes we produce these large files "just in case" when what is really needed is more careful verification of important errors and their causes. This comes up in e.g. https://github.com/elastic/elastic-agent/issues/2852 -- we might benefit from making the diagnostics smaller but more targeted, and preserving better records when we encounter errors.

Comment (cmacknz) [reply internal/pkg/agent/application/coordinator/diagnostics_test.go:1]: 99% of the time I am only looking at pre-config.yaml, state.yaml, and the logs. The computed/expected/actual only get looked at when variable substitutions or conditions are involved, which is almost always for standalone agents and usually agents on K8S. 

I would hesitate to remove them, but they definitely aren't used frequently. If we can figure out how to output those files as actual YAML matching the policy format instead of the crazyily nested google.protobuf.Struct fields that would be a nice improvement.

Comment (faec) [reply internal/pkg/agent/application/coordinator/diagnostics_test.go:1]: 👍 I think part of the problem there is that we use those protobufs not just for API calls but as our core internal representation of that data... I'm not sure how heavily we depend on it though, maybe there is a way to avoid it, or at least to work around it when encoding the most annoying cases. I will try and poke at it when I'm revisiting the error state reporting next sprint.

**Comment by faec** [comment]
As of this morning the integration tests are still failing at head (https://buildkite.com/elastic/elastic-agent/builds/1278#0188f77a-867b-4072-9107-8d3bd3fa9ce1) despite the rest of the build being fixed. Merging anyway after checking with @cmacknz and we'll keep an eye on the status as the unrelated errors in the integration tests are fixed.

**Pull request merged**
